Amend triggered action request with S2S header

KarolJastrzebski opened the pull request:
https://tools.hmcts.net/jira/browse/RPE-173

Job Scheduler needs to authenticate itself to action endpoint. It does
that by generating S2S token and adding it to the request headers.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

KarolJastrzebski committed:
Amend triggered action request with S2S header

Job Scheduler needs to authenticate itself to action endpoint. It does
that by generating S2S token and adding it to the request headers.

rkondratowicz commented on src/main/resources/application.yaml:
```diff
@@ -8,6 +8,12 @@ management:
+  s2s-auth:
+    url: ${S2S_URL}
+    totp_secret: ${TOTP_SECRET}
+    microservice: jobscheduler
```
Not sure about the current convention (worth checking ansible scripts), it may use dashes.

rkondratowicz commented on src/main/java/uk/gov/hmcts/reform/jobscheduler/config/ServiceTokenGeneratorConfiguration.java:
```diff
@@ -0,0 +1,34 @@
+        @Qualifier("serviceAuthTokenGenerator") final AuthTokenGenerator serviceAuthTokenGenerator,
+        @Value("${idam.s2s-auth.tokenTimeToLiveInSeconds:14400}") final int ttl) {
+        return new CachedServiceAuthTokenGenerator(serviceAuthTokenGenerator, ttl);
+    }
```
do we need both versions?

KarolJastrzebski replied:
Technically only one, but I would leave it as it is because of clear composition.

rkondratowicz replied:
Having two beans of the same type, when only one implementation is used only adds confusion imho.

lgonczar commented on src/main/java/uk/gov/hmcts/reform/jobscheduler/config/ServiceTokenGeneratorConfiguration.java:
```diff
@@ -0,0 +1,34 @@
+    @Bean
+    public AuthTokenGenerator cachedServiceAuthTokenGenerator(
+        @Qualifier("serviceAuthTokenGenerator") final AuthTokenGenerator serviceAuthTokenGenerator,
+        @Value("${idam.s2s-auth.tokenTimeToLiveInSeconds:14400}") final int ttl) {
```
It would be best to have the default value (14400) specified in the config file, so that we have all configuration in one place.

lgonczar replied:
Best to remove the default value from here now.

rkondratowicz commented on src/main/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJob.java:
```diff
@@ -19,10 +21,16 @@
+    public HttpCallJob(
+        RestTemplate restTemplate,
+        ActionExtractor actionExtractor,
+        @Qualifier("cachedServiceAuthTokenGenerator") AuthTokenGenerator tokenGenerator
```
What if the token is no longer valid? We'll wait for 4h until a new one is retrieved by this generator.
Perhaps on 401/403 response from the target server we could retrieve fresh token and try again?

KarolJastrzebski replied:
I can see two options:
1. lowering ttl in configuration
2. leaving it to the retry mechanism (from another story)

lgonczar replied:
Given how CachedServiceAuthTokenGenerator is implemented (no resetting), I'd just lower ttl for now, so that another attempt to execute the same job (after its failure) would use a different token. As Rafał mentioned to me, he had analysed the generator's code and discovered that some improvements can probably be made to it. The generator could check the expiry date of the tokens and request new ones if necessary. This way we (and everyone else) wouldn't have to worry about caching expired tokens.

rkondratowicz commented on src/test/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJobTest.java:
```diff
@@ -49,21 +53,25 @@ public void execute_should_send_http_request_specified_in_job_details() {
             .willReturn(newJob(HttpCallJob.class).withIdentity("id", "group").build());
 
+        Map<String, String> headers = new HashMap<>();
+        headers.put("ServiceAuthorization", "some-token");
```
What happens to this header sent by the client? does it get replaced?
I think it would be good to use some other generic header here and have a separate test (with nice name) that tests what happens in such a case.

KarolJastrzebski replied:
Yes, `ServiceAuthorization` gets replaced. I'll add the test with another header.

rkondratowicz commented on src/main/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJob.java:
```diff
@@ -32,6 +40,8 @@ public void execute(JobExecutionContext context) {
 
         HttpAction action = actionExtractor.extract(context);
 
+        action.headers.put("ServiceAuthorization", tokenGenerator.generate());
```
Let's keep the action scheduled by the client in its original state (instead of mutating it) and add this header when building a http request for RestTemplate.

lgonczar replied:
This is necessary because we can't persist the token for security reasons. Even if it was encrypted, it could expire by the time the action is executed.

lgonczar replied:
Sorry, I missed the fact this is where we execute the request already, not schedule the job. Ignore the previous comment then. Still, better not to mix these two things - HttpAction object should just represent what the client sent.

KarolJastrzebski committed:
Add tokenTimeToLiveInSeconds to the config file

KarolJastrzebski committed:
Allow setting microservice via S2S_NAME env

KarolJastrzebski committed:
Treat HttpAction headers as invariants

KarolJastrzebski committed:
Add a test for custom header

Action headers should be passed along to the endpoint.

rkondratowicz commented on src/main/resources/application.yaml:
```diff
@@ -8,6 +8,13 @@ management:
+    url: ${S2S_URL}
+    totp_secret: ${TOTP_SECRET}
+    microservice: ${S2S_NAME:jobscheduler}
+    tokenTimeToLiveInSeconds: 14400
```
allow modification via env vars

rkondratowicz commented on src/test/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJobTest.java:
```diff
@@ -49,21 +53,27 @@ public void execute_should_send_http_request_specified_in_job_details() {
             postRequestedFor(urlEqualTo("/hello-world"))
-                .withHeader("ServiceAuthorization", equalTo("some-token"))
+                .withHeader("X-Custom-Header", equalTo("anything"))
+                .withHeader("ServiceAuthorization", equalTo("newly-generated-token"))
```
Now this single test is not only about 'should_send_http_request_specified_in_job_details', but also:
- `ServiceAuthorization` header is added
- `ServiceAuthorization` header if replaced if it already existed in action specified in job details

Can we have multiple tests? If not, perhaps just update the test method name so that it's clear what it's actually testing

KarolJastrzebski replied:
Thanks, renamed the test.

rkondratowicz replied:
You've only removed 'execute_' from test name (which was the name of the method that's being tested)

lgonczar replied:
I agree, it would be better if there were tests focused on auth header alone (and not caring about other details of the request). This way we would have very granular and distinctive cases - adds auth header when absent, replaces when present.

KarolJastrzebski replied:
Refactored to more descriptive test names.

KarolJastrzebski committed:
Rename test to match assertions

rkondratowicz commented on src/test/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJobTest.java:
```diff
@@ -49,21 +53,27 @@ public void execute_should_send_http_request_specified_in_job_details() {
                 "http://localhost:8080/hello-world",
                 HttpMethod.POST,
-                ImmutableMap.of("ServiceAuthorization", "some-token"),
+                headers,
```
this can still be an immutable map with 2 headers

lgonczar replied:
`headers` as a variable is ok, though.

KarolJastrzebski replied:
done

KarolJastrzebski committed:
HttpCallJobTest: make headers immutable

lgonczar commented on src/test/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJobTest.java:
```diff
@@ -49,21 +55,27 @@ public void execute_should_send_http_request_specified_in_job_details() {
                 "http://localhost:8080/hello-world",
                 HttpMethod.POST,
-                ImmutableMap.of("ServiceAuthorization", "some-token"),
+                ImmutableMap.copyOf(headers),
```
I think we misunderstood each other. I guess what Rafał, seeing that you moved from ImmutableMap to HashMap, wanted to say is that it's still possible to do the same with ImmutableMap:

```
ImmutableMap.of(
    "X-Custom-Header", "anything",
    "ServiceAuthorization", "some-token"
)
```

This line was ok as it was (without `ImmutableMap.copyOf`), and I suppose the way headers Map is defined is also fine (although ImmutableMap is even better).

KarolJastrzebski replied:
I've refactored the tests.

KarolJastrzebski committed:
Refactor HttpCallJobTest: better naming of test methods

KarolJastrzebski committed:
Remove default value from ttl config

KarolJastrzebski committed:
Allow cache ttl modification with S2S_TTL env

KarolJastrzebski committed:
Fix typo in test name

KarolJastrzebski committed:
Use immutable maps instead of HashMap

KarolJastrzebski committed:
Checkstyle: fix import order in test

lgonczar commented on src/test/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJobTest.java:
```diff
@@ -49,22 +53,93 @@ public void execute_should_send_http_request_specified_in_job_details() {
+        );
+    }
+
+    private void actionHadHeadersSetTo(Map<String, String> headers) {
```
I see that the intention is to make given-when-then read like a proper sentence, but by convention method names should be imperative, i.e. `setActionHeadersTo`

lgonczar commented on src/test/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJobTest.java:
```diff
@@ -49,22 +53,93 @@ public void execute_should_send_http_request_specified_in_job_details() {
+        );
+    }
+
+    private void actionHadHeadersSetTo(Map<String, String> headers) {
```
I see that the intention is to make given-when-then read like a proper sentence, but by convention method names should be imperative, i.e. `setActionHeadersTo`, `executeHttpCallJob`.

Maybe we could have more versions of this method, that would allow us to set url and body?

KarolJastrzebski replied:
I have extracted literal values into constants so that they could be referenced from any test, stub, and assertion.

lgonczar commented on src/test/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJobTest.java:
```diff
@@ -49,22 +53,93 @@ public void execute_should_send_http_request_specified_in_job_details() {
+
+        // then
+        verify(
+            postRequestedFor(urlEqualTo("/hello-world"))
```
When we verify those values (`"/hello-world"`, `"some-body"`, `"newly-generated-token"`), we should also specify them in the test, so that we would know where they come from. You're already doing it for the headers.

This applies to all tests.

KarolJastrzebski replied:
I have extracted literal values into class constants so that they could be referenced from any stub, test, and assertion.

lgonczar commented on src/test/java/uk/gov/hmcts/reform/jobscheduler/jobs/HttpCallJobTest.java:
```diff
@@ -49,22 +53,93 @@ public void execute_should_send_http_request_specified_in_job_details() {
+    }
+
+    @Test
+    public void execute_calls_given_endpoint_url() {
```
I like the granularity of these tests 👍

KarolJastrzebski committed:
Replace literal values with class constants

They are to be shared between tests, mocks, and assertions.

KarolJastrzebski committed:
Rename TOTP_SECRET to S2S_SECRET

To keep consistency with other variables.

KarolJastrzebski committed:
Remove totp_ prefix from variables

Seems unnecessary

lgonczar commented:
Worth squashing

KarolJastrzebski committed:
Accept S2S_ vars in docker

KarolJastrzebski committed:
Simplify token generator configuration

Use @Bean annotated methods only when such objects are used in the code
and not for resolving configuration dependencies.
